lib/vector-core/src/transform/mod.rs

```diff
@@ -246,31 +270,61 @@ impl TransformOutputs {
         buf: &mut TransformOutputsBuf,
     ) -> Result<(), Box<dyn error::Error + Send + Sync>> {
         if let Some(primary) = self.primary_output.as_mut() {
-            let count = buf.primary_buffer.as_ref().map_or(0, OutputBuffer::len);
-            let byte_size = buf.primary_buffer.as_ref().map_or(
-                JsonSize::new(0),
-                EstimatedJsonEncodedSizeOf::estimated_json_encoded_size_of,
-            );
-            buf.primary_buffer
-                .as_mut()
-                .expect("mismatched outputs")
-                .send(&mut primary.fanout)
-                .await?;
-            primary.events_sent.emit(CountByteSize(count, byte_size));
+            let buf = buf.primary_buffer.as_mut().expect("mismatched outputs");
+            Self::send_single_buffer(buf, primary).await?;
         }
-
         for (key, buf) in &mut buf.named_buffers {
-            let count = buf.len();
-            let byte_size = buf.estimated_json_encoded_size_of();
             let output = self.named_outputs.get_mut(key).expect("unknown output");
-            buf.send(&mut output.fanout).await?;
-            output.events_sent.emit(CountByteSize(count, byte_size));
+            Self::send_single_buffer(buf, output).await?;
         }
+        Ok(())
+    }
 
+    async fn send_single_buffer(
+        buf: &mut OutputBuffer,
+        output: &mut TransformOutput,
+    ) -> Result<(), Box<dyn error::Error + Send + Sync>> {
+        for event in buf.events_mut() {
+            update_runtime_schema_definition(
+                event,
+                &output.output_id,
+                &output.log_schema_definitions,
+            );
+        }
+        let count = buf.len();
+        let byte_size = buf.estimated_json_encoded_size_of();
+        buf.send(&mut output.fanout).await?;
+        output.events_sent.emit(CountByteSize(count, byte_size));
         Ok(())
     }
 }
 
+#[allow(clippy::implicit_hasher)]
```

Comment:
It might be an interesting optimization to look into, but since we aren't actually using anything other than the default hasher now, I don't think it's worth supporting it. I'd want it to be a separate PR regardless.